src/GasEstimator.ts

```diff
@@ -0,0 +1,262 @@
+ * @returns gas estimation from the transfer
+ */
+export const estimateMaxPossibleGasTokenTransfer = async (
+    contractInteractor: ContractInteractor,
+    { request, relayData }: RelayRequest | DeployRequest
+): Promise<BigNumber> => {
+    const deploy = isDeployRequest(request);
+    let tokenEstimation: BigNumber;
+    const tokenGas = BigNumber(request.tokenGas);
+    if (tokenGas.gt(0)) {
+        tokenEstimation = BigNumber(request.tokenGas);
+    } else {
+        const erc20: ERC20Token = await contractInteractor.getERC20Token(
+            request.tokenContract
+        );
+
+        const methodToEstimate = await erc20.instance.contract.methods.transfer(
+            relayData.feesReceiver,
+            toWei('1')
+        );
+
+        let caller = relayData.callForwarder;
+        if (deploy) {
+            const { from, recoverer, index } = request as DeployRequestStruct;
+            caller = await contractInteractor.getSmartWalletAddress(
+                caller,
+                from,
+                recoverer,
+                index
+            );
+        }
+
+        const estimation = await methodToEstimate.estimateGas({
+            from: caller,
+            gasPrice: relayData.gasPrice
+        });
+
+        tokenEstimation = applyInternalCorrection(estimation);
+    }
+
+    if (tokenEstimation.isZero() && deploy) {
+        return SUBSIDY;
+    }
+
+    return applyGasCorrectionFactor(tokenEstimation);
+};
+
+/**
+ * Estimates the max possible gas consumed by executing a function
+ * @param contractInteractor object containing the contractInteractor
+ * @param estimate params to execute the estimateGas function
+ * @returns gas estimation from the execution
+ */
+export const estimateMaxPossibleGasExecution = async (
+    contractInteractor: ContractInteractor,
+    estimate: EstimateGasParams
+): Promise<BigNumber> => {
+    const estimation = await contractInteractor.estimateGas(estimate);
+    return BigNumber(estimation);
+};
+
+/**
+ * Applies the correction from internal calls
+ * @param estimation BigNumber gas estimation that needs to be corrected
+ * @returns gas estimation with the correction done
+ */
+export const applyInternalCorrection = (
+    estimation: BigNumber | number | string
+): BigNumber => {
+    const bigValue = BigNumber(estimation);
+    const correction = BigNumber(
+        constants.INTERNAL_TRANSACTION_ESTIMATE_CORRECTION
+    );
+    if (bigValue.gte(correction)) {
+        return bigValue.minus(correction);
+    }
+    return bigValue;
+};
+
+/**
+ * Applies the correction from the RSK node misscalculation
+ * @param estimation BigNumber gas estimation that needs to be corrected
+ * @returns gas estimation with the correction done
+ */
+export const applyGasCorrectionFactor = (
+    estimation: BigNumber | number | string
+): BigNumber => {
+    const bigValue = BigNumber(estimation);
+    if (constants.ESTIMATED_GAS_CORRECTION_FACTOR !== 1) {
```

Comment:
again, out-of-scope I guess but should prob be configurable

Reply:
its configurable isn't, it?

Reply:
is it? https://github.com/rsksmart/rif-relay-common/blob/f8e4ce8d90a979f3eec55a0eb5dbd5cc742c40e3/src/Constants.ts#L26

Reply:
not sure if this should be configurable, we may need to investigate more on those parameters